### PR TITLE
WIP: Reading Responses from a stream

### PR DESCRIPTION
**Pull request by hendrikebbers**
A first idea how we can read several responses from a channel. This should not be merged since all other classes are not refactored and the project will not compile. This is just to discuss the new functionality. If the method is not clear I can add some doc :)

### CODE REVIEW

**Comment by Oliver-Loeffler** [comment]
I like the idea of using the Consumer<HttpResponse> interface.  Why limiting the bytesBuffer to 48 elements? Is that just an arbitrary size?

**Comment by hendrikebbers** [comment]
48 is a random number. Do not know what the best size will be :)

File: src/main/java/net/raumzeitfalle/niohttp/HttpResponseReader.java [hunk 1]
```diff
@@ -1,33 +1,64 @@
...
-		byte[] bytes = new byte[receive.capacity()];
-		bytes[0] = receive.get();
-		for (int i = 1; i < bytes.length; i++) {
-			bytes[i] = receive.get();
-			if (!receive.hasRemaining())
-			break;
-			if (bytes[i - 1] == CRLF && bytes[i] == CRLF)
-			break;
-		}
+    public void readFromBuffer(final Consumer<HttpResponse> responseConsumer) throws IOException {
```
Comment (Oliver-Loeffler) [inline src/main/java/net/raumzeitfalle/niohttp/HttpResponseReader.java]: What is the benefit in taking a consumer versus returning a HttpResponse directly? Well, I can imagine scenarios for both approaches, I'm just curious?

**Comment by hendrikebbers** [comment]
The method reads from the channel until it is closed. Since HTTP can hold a connection / channel instead of creating a new connection for each request several requests / responses can be send over one channel. Therefore I added a "while" to the method and the method will be executed until the channel is closed. In this case the consumer will be called several times for each response that was transported.

Von meinem iPhone gesendet

**Comment by Oliver-Loeffler** [comment]
Oh, it looks as I've got the intention wrong. So the idea is to keep reading from channel as long as it is open and basically read all responses which may be there?  If so, then I'm going to revert some of the changes in HttpResponseReader. 
Well, my misunderstanding also causes sometimes the nice effect that payload is only read partially.

**Comment by Oliver-Loeffler** [comment]
HttpResponseReader now keeps reading as long as the channel is open. Here HttpResponse.fromBytes() fails as this kind of continuously reading a stream of messages is not yet implemented herein.

**Comment by Oliver-Loeffler** [comment]
Well, I saw that the consumer will be called several times per response. I do not understand why to call the consumer multiple times with one response. Except, when the consumer is supposed to build a response object, possibly. 
I also found that wen looping while channel.isOpen() this can last pretty long even with receiving no messages at all. So right now the while (channel.isOpen()) is commented out.

**Comment by Oliver-Loeffler** [comment]
The original proposal still exists in [response-reader-orig-proposal](/Oliver-Loeffler/NIOHttp/tree/response-reader-orig-proposal) and I'll have a look at it again tonight or tomorrow (considering all learnings :-)).

**Pull request merged**